picard/util/__init__.py

```diff
@@ -293,16 +296,20 @@ def sanitize_date(datestr):
     return ("", "%04d", "%04d-%02d", "%04d-%02d-%02d")[len(date)] % tuple(date)
 
 
-_re_win32_incompat = re.compile(r'["*:<>?|]', re.UNICODE)
-def replace_win32_incompat(string, repl="_"):  # noqa: E302
+def replace_win32_incompat(string, repl="_", replacements=None):  # noqa: E302
     """Replace win32 filename incompatible characters from ``string`` by
        ``repl``."""
     # Don't replace : with _ for windows drive
     if IS_WIN and os.path.isabs(string):
-        drive, rest = ntpath.splitdrive(string)
-        return drive + _re_win32_incompat.sub(repl, rest)
+        drive, string = ntpath.splitdrive(string)
     else:
-        return _re_win32_incompat.sub(repl, string)
+        drive = ''
+
+    replacements = defaultdict(lambda: repl, replacements or {})
+    for char in {'"', '*', ':', '<', '>', '?', '|'}:
+        string = string.replace(char, replacements[char])
```

Comment:
In Python3, there are `str.translate()` and `str.maketrans()`, but not sure if it would be faster and/or more convenient.
Also I think testing the presence of  `char` in `string` before calling `str.replace()` is faster in some cases .

```suggestion
        if char in string:
            string = string.replace(char, replacements[char])
```

In our case, it is unlikely we replace all chars I guess.

A quick performance test (far from complete):


```python
import timeit


s_occurences = {
    0: 'abcdefghijklmnop',
    2: 'abcd"fghijk*mnop',
    4: 'ab:d"fghijk*mn<p',
    6: 'ab:d"fg?i>k*mn<p',
    8: 'a*:d"fg?i>k*mn<:',
    'max': ':*::"::?:>:*::<:',
}


def replA(string):
    for char in {'"', '*', ':', '<', '>', '?', '|'}:
        string = string.replace(char, '_')
    return string


def replB(string):
    for char in {'"', '*', ':', '<', '>', '?', '|'}:
        if char in string:
            string = string.replace(char, '_')
    return string


trans = str.maketrans({'"': '_', '*': '_', ':': '_', '<': '_', '>': '_', '?': '_', '|': '_'})


def replC(string):
    return string.translate(trans)


for k, v in s_occurences.items():
    print(k)
    print("A: %f" % timeit.timeit(stmt=lambda: replA(s_occurences[k]), number=1000000))
    print("B: %f" % timeit.timeit(stmt=lambda: replB(s_occurences[k]), number=1000000))
    print("C: %f" % timeit.timeit(stmt=lambda: replC(s_occurences[k]), number=1000000))
```

```
0
A: 0.376292
B: 0.255101
C: 0.672888
2
A: 0.441930
B: 0.375825
C: 0.630128
4
A: 0.476545
B: 0.476716
C: 0.584534
6
A: 0.500308
B: 0.598776
C: 0.557867
8
A: 0.510370
B: 0.588787
C: 0.484488
max
A: 0.513036
B: 0.590188
C: 0.236326
```

That's tricky, because performance depends a lot on number of occurrences of the searched characters in the original string. The `translate` approach beats everything if the source string is made only of the searched characters (which is unlikely), but else is twice slower.

The extra test is improving performance when the string doesn't have much characters to replace but when there's a lot it is a tad slower.

Reply:
Given the numbers I think we could go with B then. That we have no or only a few of those characters seems to be the most common scenario for our use case. It is a bit counterintuitive why this is faster, but if it is like this it makes sense to use it.

That `str.translate` is so slow is disappointing, and your timing does not even include the `maketrans` call (which in the current architecture would need to be done each time as well, but that probably could be refactored).